controller/admin_controller.php

```diff
@@ -0,0 +1,294 @@
+	 * @param \phpbb\request\request            $request              Request object
+	 * @param \phpbb\template\template          $template             Template object
+	 * @param \phpbb\user                       $user                 User object
+	 * @param string                            $root_path            phpBB root path
+	 * @param string                            $php_ext              php_ext
+	 * @access public
+	 */
+	public function __construct(\phpbb\config\config $config, \phpbb\db\driver\driver_interface $db, \phpbb\language\language $language, \phpbb\log\log $log, \phpbb\request\request $request, \phpbb\template\template $template, \phpbb\user $user, $root_path, $php_ext)
+	{
+		$this->config = $config;
+		$this->db = $db;
+		$this->language = $language;
+		$this->log = $log;
+		$this->request = $request;
+		$this->template = $template;
+		$this->user = $user;
+		$this->phpbb_root_path = $root_path;
+		$this->php_ext = $php_ext;
+	}
+
+	/**
+	 * Display the options a user can configure for this extension
+	 *
+	 * @return null
+	 * @access public
+	 */
+	public function display_options()
+	{
+		$this->template->assign_vars(array(
+			'IDEAS_POSTER'		=> $this->get_ideas_topics_poster_username(),
+			'IDEAS_BASE_URL'	=> ($this->config['ideas_base_url']) ?: '',
+
+			'S_FORUM_SELECT_BOX'	=> $this->select_ideas_forum(),
+			'S_IDEAS_FORUM_ID'		=> !empty($this->config['ideas_forum_id']),
+
+			'U_ACTION'			=> $this->u_action,
+			'U_FIND_USERNAME'	=> append_sid("{$this->phpbb_root_path}memberlist.{$this->php_ext}", 'mode=searchuser&amp;form=acp_phpbb_ideas_settings&amp;field=ideas_poster_id&amp;select_single=true'),
+		));
+	}
+
+	/**
+	 * Get Ideas poster bot user ID
+	 *
+	 * @return int user_id Ideas bot user ID
+	 * @access protected
+	 */
+	protected function get_ideas_topics_poster_id()
+	{
+		$sql = 'SELECT user_id
+			FROM ' . USERS_TABLE . "
+			WHERE username_clean = '" . $this->db->sql_escape(utf8_clean_string($this->cfg_array['ideas_poster_id'])) . "'";
+		$result = $this->db->sql_query($sql);
+		$user_id = (int) $this->db->sql_fetchfield('user_id');
+		$this->db->sql_freeresult($result);
+
+		return $user_id;
+	}
+
+	/**
+	 * Get Ideas poster bot username
+	 *
+	 * @return string Ideas bot username
+	 * @access protected
+	 */
+	protected function get_ideas_topics_poster_username()
+	{
+		$sql = 'SELECT username FROM ' . USERS_TABLE . '
+			WHERE user_id = ' . (int) $this->config['ideas_poster_id'];
+		$this->db->sql_query($sql);
+		$username = $this->db->sql_fetchfield('username');
+
+		return ($username !== false) ? $username : '';
+	}
+
+	/**
+	 * Set configuration options
+	 *
+	 * @return null
+	 * @access public
+	 */
+	public function set_config_options()
+	{
+		$errors = array();
+
+		// This method is called on submit, so set flag to true initially
+		$submit = true;
+
+		$this->cfg_array = ($this->request->is_set('config')) ? $this->request->variable('config', array('' => ''), true) : $this->config;
```

Comment:
Is this something we still have to be doing? I just not thrilled that we are duplicating the entire $config object.